src/plugins/embeddable/public/lib/embeddables/attribute_service.ts

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { SavedObjectsClientContract } from '../../../../../core/public';
+import {
+  SavedObjectEmbeddableInput,
+  isSavedObjectEmbeddableInput,
+  EmbeddableInput,
+  IEmbeddable,
+} from '.';
+import { SimpleSavedObject } from '../../../../../core/public';
+
+export class AttributeService<
+  SavedObjectAttributes,
+  ValType extends EmbeddableInput & { attributes: SavedObjectAttributes },
+  RefType extends SavedObjectEmbeddableInput
+> {
+  constructor(private type: string, private savedObjectsClient: SavedObjectsClientContract) {}
+
+  public async unwrapAttributes(input: RefType | ValType): Promise<SavedObjectAttributes> {
+    if (isSavedObjectEmbeddableInput(input)) {
+      const savedObject: SimpleSavedObject<SavedObjectAttributes> = await this.savedObjectsClient.get<
```

Comment:
we are using saved object client to get a saved object .... our general approach in kibana is to actually use a higher level abstraction (savedVisualizationLoader, savedSearchLoader, indexPatternsService) which will usually perform quite some magic on top of getting the saved object.

i am wondering do we/will we have any embeddable where saved object attributes are gonna exactly equal embeddable input ?

Reply:
As discussed in our sync -- just documenting here -- The attribute service is meant to be used as a higher level abstraction in multiple places including, lens & visualize to provide a standardized pattern for using Embeddables by reference or value.